src/test/java/jenkins/plugins/slack/webhook/WebhookEndpointTest.java

```diff
@@ -0,0 +1,208 @@
+
+    @Test
+    public void testUnconfiguredSlackToken() throws Exception {
+        HtmlForm form = jenkinsRule.createWebClient().goTo("configure").getFormByName("config");
+        form.getInputByName("_.slackOutgoingWebhookURL").setValueAttribute(URL);
+        jenkinsRule.submit(form);
+
+        WebResponse response = makeRequest(null);
+
+        assertThat(response.getStatusCode(), is(HTTP_OK));
+        assertThat(getSlackMessage(response).getText(), is("Slack token not set"));
+    }
+
+    @Test
+    public void testNoTextPostData() throws Exception {
+        setConfigSettings();
+        List<NameValuePair> goodToken = new ArrayList<NameValuePair>();
+        goodToken.add(new NameValuePair("token", "GOOD_TOKEN"));
+
+        WebResponse response = makeRequest(goodToken);
+        assertThat(getSlackMessage(response).getText(), is("Invalid command, text field required"));
+    } 
+
+    @Test
+    public void testNoTriggerWordPostData() throws Exception {
+        setConfigSettings();
+        List<NameValuePair> goodToken = new ArrayList<NameValuePair>();
+        goodToken.add(new NameValuePair("token", "GOOD_TOKEN"));
+        goodToken.add(new NameValuePair("text", "jenkins"));
+
+        WebResponse response = makeRequest(goodToken);
+        assertThat(getSlackMessage(response).getText(), is("Invalid command, trigger_word field required"));
+    }
+
+    @Test
+    public void testInvalidConfiguredSlackToken() throws Exception {
+        GlobalConfig config = GlobalConfiguration.all().get(GlobalConfig.class);        
+        assertThat(config.getSlackOutgoingWebhookToken(), is(nullValue()));
+
+        setConfigSettings();
+
+        List<NameValuePair> badData = new ArrayList<NameValuePair>();
+        badData.add(new NameValuePair("token", "BAD_TOKEN"));
+
+        WebResponse response = makeRequest(badData);
+        
+        assertThat(getSlackMessage(response).getText(), is("Invalid Slack token"));
+    }
+
+    @Test
+    public void testInvalidTriggerWord() throws Exception {
+        setConfigSettings();
+        data.add(new NameValuePair("text", "jenkinns list projects"));
+        WebResponse response = makeRequest(data);
+        assertThat(getSlackMessage(response).getText(), is("Invalid command, invalid trigger_word"));
+    }
+
+    @Test
+    public void testListProjects() throws Exception {
+        setConfigSettings(); 
+        data.add(new NameValuePair("text", "jenkins list projects")); 
+        WebResponse response = makeRequest(data);
+        assertThat(getSlackMessage(response).getText(), is("*Projects:*\n>_No projects found_"));
+
+        FreeStyleProject project = jenkinsRule.createFreeStyleProject(LONG_PROJECT_NAME);
+        Thread.sleep(30);
+        response = makeRequest(data);
+        assertThat(getSlackMessage(response).getText(), is("*Projects:*\n>*"+LONG_PROJECT_NAME+"*\n>*Last Build:* #TBD\n>*Status:* TBD\n\n\n"));
+
+        project.scheduleBuild2(0);
+        Thread.sleep(200);
```

Comment:
I was able to remove all the sleeps in these tests, scheduling builds with `project.scheduleBuild2(0).get()`, and everything is still passing for me. That shaved around 9 seconds off the build time on my machine.